### PR TITLE
Forcing API refactor

### PR DESCRIPTION
**Pull request by glwagner**
This PR changes the name of the `Forcing` constructor to `ModelForcing` to clarify the difference between constructors for forcing objects on single fields, and constructors for the `NamedTuple` of forcing objects for all model fields. It also conventionalizes the name of the fields in `ModelForcing`, so that 

```julia
forcing = ModelForcing(u=u_forcing)
```

is used to apply the forcing function `u_forcing` to `u`. 

In addition it adds a new type "`SimpleForcing`", which permits users to pass functions of `x`, `y`, `z`, and `t`. A few validating lines in the `ModelForcing` constructor ensure that forcing functions for `u`, `v`, and `w` are applied at the correct locations.

For example, we can now write

```julia
T_forcing = SimpleForcing((x, y, z, t) -> exp(z) * cos(t))

model = Model(forcing=ModelForcing(T=T_forcing))
```
To create a forcing on temperature that decays exponentially in z and oscillates in time.

To test this functionality, this PR introduces an example "`ocean_convection_with_plankton`" that uses `SimpleForcing` to model the growth and decay of a plankton-like tracer.

### CODE REVIEW

File: src/forcing.jl [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+"""
+struct SimpleForcing{Lx, Ly, Lz, F}
+    func :: F
+end
+
+"""
+    SimpleForcing([location=(Cell, Cell, Cell),] forcing::SimpleForcing)
+
+Construct forcing for a field at `location` using `forcing.func`.
+"""
```
Comment (ali-ramadhan) [inline src/forcing.jl]: I might add an example to the docstring like the one from the PR comment
```julia
T_forcing = SimpleForcing((x, y, z, t) -> exp(z) * cos(t))
```
or for people who don't know about anonymous functions
```julia
FT(x, y, z, t) = exp(z) * cos(t)
T_forcing = SimpleForcing(FT)
```

File: src/forcing.jl [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+    @inbounds forcing.func(xnode(X, i, grid), ynode(Y, j, grid), znode(Z, k, grid), time)
+
+at_location(location, u::Function) = u
+at_location(location, u::SimpleForcing) = SimpleForcing(location, u)
+
+"""
+    ModelForcing(; kwargs...)
+
+Return a named tuple of forcing functions for each solution field.
+"""
```
Comment (ali-ramadhan) [inline src/forcing.jl]: Might also add a docstring examples here like the one from the PR comment
```julia
T_forcing = SimpleForcing((x, y, z, t) -> exp(z) * cos(t))
model = Model(forcing=ModelForcing(T=T_forcing))
```

File: examples/ocean_convection_with_plankton.jl [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+
+    sca(axs[1]); cla()
+    pcolormesh(xC, zF, data(model.velocities.w)[:, 1, :])
+    title("Vertical velocity")
+    xlabel("\$ x \$ (m)")
+    ylabel("\$ z \$ (m)")
+
+    sca(axs[2]); cla()
+    pcolormesh(xC, zC, data(model.tracers.S)[:, 1, :])
+    title("Phytoplankton concentration")
```
Comment (ali-ramadhan) [inline examples/ocean_convection_with_plankton.jl]: Salty phytoplankton!

Comment (glwagner) [reply examples/ocean_convection_with_plankton.jl]: Heh. With arbitrary tracers we can just call them :plankton !

File: src/forcing.jl (hunk 2 above)
Comment (glwagner) [reply src/forcing.jl]: Easy

**Comment by glwagner** [comment]
We can extend SimpleForcing with a field .parameters to optionally admit parameters via dispatch, eg

u_forcing = SimpleForcing((x, y, z, t, parameters) -> #something, parameters=user_defined_parameters)

What do you think?

Interestingly, putting parameters in the forcing object obviates the need for model.parameters. Ultimately this design may be preferred; I suppose we will see.

**Comment by ali-ramadhan** [comment]
Hmmm, yeah I guess `model.parameters` may be useful in boundary conditions (possibly elsewhere too) but if we only end up using them for forcing functions then that makes sense.

**Comment by glwagner** [comment]
> Hmmm, yeah I guess model.parameters may be useful in boundary conditions (possibly elsewhere too)

To make `model.parameters` fully obsolete, we would need wrappers / interfaces for boundary conditions; eg

```julia
bc = SimpleBoundaryCondition(Flux, (x, y, t, parameters) -> something, parameters=parameters)
```

Not suggesting we do it now, but something to think about for the future.

**Pull request merged**